src/render/ecs_system.cpp

```diff
@@ -35,8 +37,45 @@ struct RenderingSystemState {
     // Also only used when on the CPU backend
     uint64_t *instanceWorldIDsCPU;
     uint64_t *viewWorldIDsCPU;
+
+    MeshBVH *bvhs;
+    uint32_t numBVHs;
+
+    bool enableRaycaster;
 };
 
+inline uint32_t leftShift3(uint32_t x)
+{
+    if (x == (1 << 10)) {
+        --x;
+    }
+
+    x = (x | (x << 16)) & 0b00000011000000000000000011111111;
+    x = (x | (x << 8)) & 0b00000011000000001111000000001111;
+    x = (x | (x << 4)) & 0b00000011000011000011000011000011;
+    x = (x | (x << 2)) & 0b00001001001001001001001001001001;
+
+    return x;
+}
+
+uint32_t encodeMorton3(const Vector3 &v) {
+    return (leftShift3(*((uint32_t *)&v.z)) << 2) | 
```

Comment:
This is illegal. use std::bit_cast. But I think this is not actually the correct way to do morton codes. Usually I see the FP values converted to some fixed point spacing and then merged together. Unless the leftShift3 is handling this correctly above? You're going to be mixing together exponents and mantissa.

Reply:
thought so too but https://stackoverflow.com/questions/26856268/morton-index-from-2d-point-with-floats